### ISSUE TITLE
[RFC] Add Gemma3n models

### ISSUE THREAD

**Issue by hashkanna** [sgl-project/sglang-jax#1078]
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

## 1. Abstract

Add Gemma 3n support to sglang-jax after core Gemma 3 text and vision support
lands. Gemma 3n should be tracked separately because it is not a small variant
of Gemma 3 Vision: it introduces Per-Layer Embeddings, AltUp/Laurel-style
blocks, shared-KV semantics, native audio, and a MobileNet/TIMM-family vision
tower decision.

This RFC covers the E2B and E4B Gemma 3n checkpoints in staged milestones:

1. Text/model bring-up and checkpoint mapping.
2. Shared-KV, PLE, and residual-block validation.
3. Precomputed multimodal embedding merge.
4. Native audio tower.
5. Raw image/video support after deciding the MobileNet/TIMM strategy.

## 2. Motivation

Gemma 3n is Google's edge-focused multimodal Gemma series, released as the E2B
and E4B variants. It is not a smaller Gemma 3 Vision: Per-Layer Embeddings (PLE)
for memory-efficient inference, AltUp/Laurel-style residual blocks, cross-layer
shared-KV semantics, native audio understanding, and a MobileNet/TIMM-family
vision tower mean Gemma 3n needs its own model classes, runtime cache behavior,
and tower decisions rather than reusing Gemma 3 Vision plumbing. Native
sglang-jax support brings edge-grade multimodal inference (text, image, video,
audio) to TPU.

The repo already recognizes `Gemma3nForConditionalGeneration` as a multimodal
architecture name in `python/sgl_jax/srt/configs/model_config.py`, but no
native Gemma 3n modules, weight mappings, runtime cache behavior, audio
support, or raw image/video tower support exist in upstream `main`. This RFC
adds them in staged milestones so that text bring-up, cache correctness,
multimodal embedding merge, audio support, and the raw image/video tower
decision can each be reviewed independently. Core Gemma 3 text and vision
support is a prerequisite for shared utilities and is tracked separately as
the Gemma 3 core RFC.

## 3. Baseline Already In Repo

| Component | Status in `main` |
|---|---|
| `Gemma3nForConditionalGeneration` listed in `multimodal_model_archs` | Present in `python/sgl_jax/srt/configs/model_config.py` |
| Multimodal staged runtime | Present per #552 (`srt/multimodal/{manager,model_executor,models}/...`) |
| Stage config registry | Present in `srt/multimodal/models/static_configs/yaml_registry.py` |
| Stage class resolver | Present in `srt/multimodal/manager/stage.py::get_model_class()` |
| Native Gemma 3n text model, PLE, AltUp/Laurel, shared-KV handling | Absent; this RFC adds them |
| Native Gemma 3n audio tower | Absent; this RFC adds it after text bring-up |
| Native Gemma 3n MobileNet/TIMM image/video tower | Open; this RFC proposes deferring or splitting this decision |

## 4. Scope And Variant Coverage

| Family | Representative checkpoints | HF architecture | Modalities | Planned support |
|---|---|---|---|---|
| Gemma 3n E2B | `google/gemma-3n-E2B`, `google/gemma-3n-E2B-it` | `Gemma3nForConditionalGeneration` | Text, image, video, audio | N1-N5 |
| Gemma 3n E4B | `google/gemma-3n-E4B`, `google/gemma-3n-E4B-it` | `Gemma3nForConditionalGeneration` | Text, image, video, audio | N1-N5 |
| Gemma 3n LiteRT preview artifacts | LiteRT/preview names where applicable | Loader-format dependent | Text, image, video, audio | Stage recognition only until loader parity is validated |

The first useful target is E2B text plus precomputed multimodal embeddings. E4B
and raw image/video support should follow after the text path and placeholder
merge are stable.

### 5.1 Text And Conditional Model

Gemma 3n should add native model classes:

```text
Gemma3nForCausalLM
  Gemma3nModel
    embed_tokens
    Gemma3nDecoderLayer x N
      attention
      PLE / per-layer embedding path
      AltUp / Laurel-style residual blocks
      MLP / feed-forward path
    final norm
  logits processor

Gemma3nForConditionalGeneration
  language model
  multimodal embedding merge hooks
```

The exact decoder implementation should follow the HF/SGLang GPU/vLLM Gemma 3n
references while preserving sglang-jax conventions for NNX modules, sharding, and
weight mappings.

### 5.2 Load-bearing Model Differences

| Area | Gemma 3n requirement | Review concern |
|---|---|---|
| Per-Layer Embeddings | Add per-layer embedding tables and projection/norm behavior | Must not be treated as ordinary token embeddings |
| AltUp / Laurel | Implement residual/up-projection blocks according to upstream config | Needs tiny-config construction tests |
| Shared KV | Layers with `num_kv_shared_layers > 0` must read K/V from earlier owner layers | Flash backend may not support same-forward cross-layer K/V reads |
| Native attention fallback | Shared-KV configs should route to native attention until flash support is proven | Avoid silent cache correctness bugs |
| Precomputed embeddings | Accept already-computed vision/audio embeddings and merge at modality placeholders | Enables a useful first multimodal slice without blocking on MobileNet/TIMM |
| Audio tower | Add native audio tower after text path is stable | Separate validation and checkpoint mapping |
| Raw image/video tower | Decide whether to port MobileNet/TIMM, depend on TIMM, or defer raw image/video | Main open architecture question |

### 5.3 Shared-KV Semantics

Gemma 3n shared-KV support should be explicit:

- A layer that shares K/V should identify the physical owner layer it reads from.
- The owner layer writes K/V to the cache.
- The sharing layer should not overwrite the owner layer's K/V for the same
  forward pass.
- Until the flash backend exposes safe same-forward cross-layer K/V reads,
  shared-KV Gemma 3n configs should use native attention for correctness.

No-download tests should cover the cache mutation behavior directly with a tiny
config rather than relying only on end-to-end checkpoint smokes.

### 5.4 Multimodal Path

The first Gemma 3n multimodal slice should avoid making raw image/video tower
support the blocker. It should support precomputed embeddings first:

```text
Tokenizer / request
  -> input_ids with vision/audio placeholder tokens
  -> optional precomputed vision_embeds / audio_embeds

Gemma3nForConditionalGeneration
  -> embeds text tokens
  -> projects or validates precomputed modality embeddings
  -> replaces modality placeholders
  -> runs Gemma 3n text stack
```

This makes E2B/E4B text and embedding merge reviewable before the native
MobileNet/TIMM tower decision is resolved.

### 5.5 Audio Tower

The native audio milestone should cover:

- audio feature input shape and padding/truncation policy;
- USM-style SSCP / conformer tower components where required by the reference
  checkpoint;
- projection into the Gemma 3n text hidden size;
- audio placeholder replacement;
- no-download tower construction and mapping tests;
- short audio smoke where checkpoint access and TPU capacity allow.

### 5.6 Image/Video Tower Decision

Gemma 3n's raw image/video path depends on a MobileNet/TIMM-family tower. There
are three viable strategies:

1. Port `mobilenetv5_300m_enc` to native JAX.
   - Cleanest runtime story.
   - Largest implementation scope.
2. Depend on Torch/TIMM for the tower only.
   - Smaller porting scope.
   - Adds a runtime dependency that sglang-jax may not want.
3. Accept precomputed vision embeddings first and defer native raw image/video
   support.
   - Smallest first Gemma 3n slice.
   - Limits out-of-the-box image/video serving.

Default proposal: use option 3 for the first Gemma 3n PR train, then open a
focused MobileNet/TIMM tower RFC for
native raw image/video
support.

## 6. Implementation Plan

The milestones are ordered by dependency and should start after core Gemma 3
text+vision support is either merged or close enough that shared utilities are
stable.

| Milestone | Scope | Files touched (new unless noted) |
|---|---|---|
| N1: Gemma 3n text skeleton | Register `Gemma3nForCausalLM` / `Gemma3nForConditionalGeneration`; tiny-config construction; base weight mappings | `srt/models/gemma3n.py`; `python/sgl_jax/test/test_gemma3n_model.py` |
| N2: PLE, AltUp/Laurel, shared KV | Implement PLE and residual blocks; route shared-KV configs to native attention; add cache correctness tests | `srt/models/gemma3n.py`; attention/cache tests |
| N3: Precomputed multimodal embeddings | Merge precomputed vision/audio embeddings at modality placeholders; validate placeholder counts and hidden sizes | `srt/models/gemma3n.py`; scheduler/input-embedding handoff if needed |
| N4: Native audio tower | Add audio tower, projection, audio placeholder merge, mapping tests, short audio smoke | `srt/multimodal/models/gemma3n/audio*.py` or model-local audio modules |
| N5: Raw image/video support | Resolve MobileNet/TIMM strategy; add raw image/video bridge or document explicit precomputed-only support | `srt/multimodal/models/gemma3n/vision*.py` if native support is chosen |

## 7. Validation Plan

No-download tests:

- Registry exposure for Gemma 3n model classes.
- Tiny-config construction for text and conditional paths.
- Weight mapping coverage for text, PLE, and modality projection weights.
- PLE / AltUp / Laurel path construction.
- Shared-KV cache mutation correctness.
- Native-attention routing for shared-KV configs.
- Precomputed vision/audio embedding merge.
- Placeholder count and hidden-size mismatch errors.
- Audio tower construction and shape tests once N4 lands.

Runtime smokes:

- E2B short text prompt on TPU.
- E4B short text prompt where capacity allows.
- Precomputed vision/audio embedding prompt smoke.
- Short audio smoke after the native audio tower lands.
- Raw image/video prompt smoke only after the MobileNet/TIMM strategy is chosen.

## 8. Risks And Open Questions

1. Shared-KV correctness is the main runtime risk. Static construction tests are
   not enough; cache writes and reads need focused tests.
2. MobileNet/TIMM ownership is unresolved. A native JAX tower is the cleanest
   long-term option, but it may deserve a separate RFC.
3. Audio support has a different validation surface from image/text support and
   should not block the first Gemma 3n text PR.
4. Gated checkpoints require HF credentials on validation machines.
5. LiteRT and alternate loader formats should not be claimed until the loader
   path is validated separately.

### Related resources

- #1069: Gemma 3 core text/vision RFC.
- #816: Gemma3-Vision feature request.
- #909: Q2 roadmap.
- SGLang GPU: `sglang/python/sglang/srt/models/gemma3n_causal.py`,
  `gemma3n_mm.py`, `gemma3n_audio.py`.
- vLLM GPU: `vllm/vllm/model_executor/models/gemma3n.py`, `gemma3n_mm.py`,
  `gemma3n_audio_utils.py`.
- vLLM TPU plugin / tpu-inference: `tpu-inference/tpu_inference/models/jax/`
  for JAX-on-TPU sharding and loader patterns where applicable.
- HF Transformers: `transformers/models/gemma3n/`.
